src/junit.rs

```diff
@@ -65,104 +66,173 @@ fn populate(rel_attrs: RelevantAttrs, testsuite: String) -> PyResult<Testrun> {
         testsuite,
         failure_message: None,
+        filename: rel_attrs.file,
     })
 }
 
 #[pyfunction]
-pub fn parse_junit_xml(file_bytes: &[u8]) -> PyResult<Vec<Testrun>> {
+pub fn parse_junit_xml(file_bytes: &[u8]) -> PyResult<ParsingInfo> {
     let mut reader = Reader::from_reader(file_bytes);
     reader.config_mut().trim_text(true);
 
-    let mut list_of_test_runs = Vec::new();
+    let mut list_of_test_runs: Vec<Testrun> = Vec::new();
     let mut saved_testrun: Option<Testrun> = None;
 
     let mut curr_testsuite = String::new();
     let mut in_failure: bool = false;
 
     let mut buf = Vec::new();
-    loop {
-        match reader.read_event_into(&mut buf) {
-            Err(e) => {
-                break Err(ParserError::new_err(format!(
-                    "Error parsing XML at position: {} {:?}",
-                    reader.buffer_position(),
-                    e
-                )))
+
+    // set when we parse the testsuites tag
+    let mut testsuites_name: String = "".to_string();
+
+    // set every time we finish parsing a testcase
+    let mut testsuite_names: Vec<String> = Vec::new();
+    let mut filenames = Vec::new();
+    let mut failure_messages = Vec::new();
+
+    // set the first time we finish parsing a testcase
+    let mut example_class_name = "".to_string();
+    let mut example_test_name = "".to_string();
+
+    let mut on_testcase_finish = |testrun: Testrun, ts: String| {
+        testsuite_names.push(ts.clone());
+        match testrun.filename.clone() {
+            Some(x) => {
+                filenames.push(x);
             }
-            Ok(Event::Eof) => {
-                break Ok(list_of_test_runs);
+            None => {}
+        }
+        match testrun.failure_message.clone() {
+            Some(x) => {
+                failure_messages.push(x);
             }
-            Ok(Event::Start(e)) => match e.name().as_ref() {
-                b"testcase" => {
-                    let rel_attrs = get_relevant_attrs(e.attributes())?;
-                    saved_testrun = Some(populate(rel_attrs, curr_testsuite.clone())?);
-                }
-                b"skipped" => {
-                    let testrun = saved_testrun
-                        .as_mut()
-                        .ok_or(ParserError::new_err("Error accessing saved testrun"))?;
-                    testrun.outcome = Outcome::Skip;
-                }
-                b"error" => {
-                    let testrun = saved_testrun
-                        .as_mut()
-                        .ok_or(ParserError::new_err("Error accessing saved testrun"))?;
-                    testrun.outcome = Outcome::Error;
+            None => {}
+        }
+        if example_class_name == "" {
+            example_class_name = testrun.classname.clone();
+        }
+        if example_test_name == "" {
+            example_test_name = (testrun.name).clone();
+        }
+        list_of_test_runs.push(testrun);
+    };
+
+    let loop_result =
```

Comment:
Instead of `break` with an `Err`, I would just `return Err` directly.
Then you can just `break` out of the loop without having to then `match` on the result down below.

As you are just breaking with `list_of_test_runs`, you can directly use that variable after the loop.